### PR TITLE
SNOW-418514 Add Typed DataFrames and Empty Tables

### PR DESCRIPTION
**Pull request by sfc-gh-jbahk**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-418514

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: test/integ/pandas/test_arrow_pandas.py [hunk 1]
```diff
@@ -1094,3 +1099,34 @@ def test_sessions_used(conn_cnx, fetch_fn_name, pass_connection):
...
+    ) as cnx:
+        with cnx.cursor() as cur:
+            cur.execute(
+                "select 1::integer, 2.3::double, 'foo'::string, current_timestamp()::timestamp where 1=0"
+            )
+            batches = cur.get_result_batches()
+            batch = batches[0].to_pandas()
+
+            assert batch.dtypes is not None
+            # assert batch.empty is not True TODO
```
Comment (sfc-gh-mkeller) [inline test/integ/pandas/test_arrow_pandas.py:1125]: Please format this string differently:
```suggestion
            # TODO: assert batch.empty is not True
```

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: src/snowflake/connector/result_batch.py [hunk 2]
```diff
@@ -651,21 +664,45 @@ def _get_arrow_iter(
         """Returns an iterator for this batch which yields a pyarrow Table"""
         return self._create_iter(iter_unit=IterUnit.TABLE_UNIT, connection=connection)
 
-    def to_arrow(
-        self, connection: Optional["SnowflakeConnection"] = None
-    ) -> Optional[Table]:
+    def _create_emtpy_table(self) -> Table:
+        """Returns emtpy Arrow table based on schema"""
+        if installed_pandas:
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/result_batch.py:669]: I don't think this is necessary inside of this function
```suggestion
```

File: test/integ/pandas/test_arrow_pandas.py [hunk 3]
```diff
@@ -1094,3 +1098,35 @@ def test_sessions_used(conn_cnx, fetch_fn_name, pass_connection):
             ) as get_session_mock:
                 fetch_fn(connection=connection)
                 assert get_session_mock.call_count == (1 if pass_connection else 0)
+
+
+def assert_dtype_equal(a, b):
+    """Pandas method of asserting the same numpy dtype of variables by computing hash"""
```
Comment (sfc-gh-mkeller) [inline test/integ/pandas/test_arrow_pandas.py:1104]:
```suggestion
    """Pandas method of asserting the same numpy dtype of variables by computing hash."""
```

**Review by sfc-gh-mkeller — APPROVED** [review]

File: src/snowflake/connector/result_batch.py (hunk 2 above)
Comment (sfc-gh-jbahk) [reply src/snowflake/connector/result_batch.py:669]: This is the safeguard to protect against the unresolved imports from pyarrow when lazy evaluating.

**Pull request merged**